src/EventStore.Core/LogV3/PartitionManager.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using EventStore.Common.Exceptions;
+using EventStore.Core.TransactionLog;
+using EventStore.Core.TransactionLog.LogRecords;
+using EventStore.LogCommon;
+using ILogger = Serilog.ILogger;
+
+namespace EventStore.Core.LogV3
+{
+	public interface IPartitionManager {
+		
+		Guid? RootId { get; }
+		Guid? RootTypeId { get; }
+		
+		void Initialize(int epochNumber);
+	}
+
+	public class PartitionManager : IPartitionManager {
+		private static readonly ILogger _log = Serilog.Log.ForContext<PartitionManager>();
+		private readonly ITransactionFileReader _reader;
+		private readonly ITransactionFileWriter _writer;
+		private readonly LogV3RecordFactory _recordFactory;
+
+		private const string RootPartitionName = "Root";
+		private const string RootPartitionTypeName = "Root";
+		
+		public Guid? RootId { get; private set; }
+		public Guid? RootTypeId  { get; private set; }
+
+		public PartitionManager(
+			ITransactionFileReader reader, ITransactionFileWriter writer, LogV3RecordFactory recordFactory) {
+			_reader = reader;
+			_writer = writer;
+			_recordFactory = recordFactory;
+		}
+
+		public void Initialize(int epochNumber) {
+			if (epochNumber == 0) {
+				CreateRootPartition();
+			} else if (epochNumber > 0 && !RootId.HasValue) {
+				ReadRootPartition();
+				
+				if (!RootId.HasValue || !RootTypeId.HasValue) {
+					throw new ApplicationInitializationException("Failed to initialize the root partition!");
+				}
+			}
+		}
+
+		private void CreateRootPartition()
+		{
+			// below code only takes into account offline truncation
+			RootTypeId = Guid.NewGuid();
+			long pos = _writer.Checkpoint.ReadNonFlushed();
+			var rootPartitionType = _recordFactory.CreatePartitionTypeRecord(
+				timeStamp: DateTime.UtcNow,
+				logPosition: pos,
+				partitionTypeId: RootTypeId.Value,
+				partitionId: Guid.Empty, 
+				name: RootPartitionTypeName);
+			
+			if (!_writer.Write(rootPartitionType, out pos))
+				throw new Exception($"Failed to write root partition type!");
+
+			_log.Debug("Root partition type created, id: " + RootTypeId);
+
+			RootId = Guid.NewGuid();
+			var rootPartition = _recordFactory.CreatePartitionRecord(
+				timeStamp: DateTime.UtcNow,
+				logPosition: pos, 
+				partitionId: RootId.Value, 
+				partitionTypeId: RootTypeId.Value, 
+				parentPartitionId: Guid.Empty, 
+				flags: 0, 
+				referenceNumber: 0, 
+				name: RootPartitionName);
+
+			if (!_writer.Write(rootPartition, out pos))
+				throw new Exception($"Failed to write root partition!");
+
+			_writer.Flush();
+
+			_recordFactory.SetRootPartitionId(RootId.Value);
+			
+			_log.Debug("Root partition created, id: " + RootId);
+		}
+
+		private void ReadRootPartition()
+		{
+			SeqReadResult result;
+			_reader.Reposition(0);
+			while ((result = _reader.TryReadNext()).Success)
```

Comment:
@pvanbuijtene it looks like we can be more strict here with what types records to expect:
Epoch(0) -> Partition type record -> Partition record

so we can read the records one by one in sequence and throw if any one of them does not match the expected record type

Reply:
(this can potentially prevent reading through the whole database in case some data is corrupted, although it's very unlikely)